Dockerfile.neuron

```diff
@@ -22,17 +25,17 @@ RUN python3 -m pip install sentencepiece transformers==4.36.2 -U
 RUN python3 -m pip install transformers-neuronx --extra-index-url=https://pip.repos.neuron.amazonaws.com -U
 RUN python3 -m pip install --pre neuronx-cc==2.12.* --extra-index-url=https://pip.repos.neuron.amazonaws.com -U
 
-COPY ./vllm /app/vllm/vllm
-COPY ./setup.py /app/vllm/setup.py
-COPY ./requirements-common.txt /app/vllm/requirements-common.txt
-COPY ./requirements-neuron.txt /app/vllm/requirements-neuron.txt
+COPY . /app/vllm
 
-RUN cd /app/vllm \
-    && python3 -m pip install -U -r requirements-neuron.txt
+# Update the build requirements' torch version, since the base image has torch==2.1.2
+RUN cd /app/vllm && \
+    sed -i 's/torch.*==.*2.4.0/torch==2.1.2/' requirements-build.txt pyproject.toml && \
```

Comment:
sorry I don't get it. why do other hardware need to use `requirements-build.txt` and do this string manipulation? they just need `requirements-common.txt` and `requirements-neuron.txt` before this pr.

Reply:
@youkaichao I'd expect `requirements-build.txt` (and synced build requirements in `pyproject.toml`) to include the modules that are required to build `vllm` for all platforms: I think it really makes it easier to mess up the build process for other platforms if we use `requirements-build.txt` for one platform and manually specify some build dependencies in the plaform-specific Dockerfiles.

One example on how this is broken follows.


Running:

```bash
VLLM_TARGET_DEVICE="xpu" pip install git+https://github.com/vllm-project/vllm
```

Tries to build vllm in an ephemeral virtualenv, in which the build dependencies specified in `pyproject.toml` are installed:

```
[build-system]
# Should be mirrored in requirements-build.txt
requires = [
    "cmake>=3.26",
    "ninja",
    "packaging",
    "setuptools>=61",
    "setuptools-scm>=8.0",
    "torch == 2.4.0",
    "wheel",
    "jinja2",
]
```
meaning this attempts to build vllm with torch 2.4.0, while xpu requires `2.1.0`, making the build fail.

Not using `requirments-build.txt` also becomes an issue whenever a new build requirement is added: we need to go through each Dockerfile and manually change the install process to include the new dependency (which is also prone to errors and de-syncing).

While using `sed` to replace the version strings in `requirements-build.txt` and `pyproject.toml` is a hack, it's just a temporary workaround that we need to make the builds a bit saner until we implement a better build backend.

In that regard, here's something I proposed a few weeks back, which I would love to receive some feedback on: https://github.com/vllm-project/vllm/pull/7525.

Reply:
is `VLLM_TARGET_DEVICE="xpu" pip install git+https://github.com/vllm-project/vllm` supported?

> whenever a new build requirement is added

isn't it just adding to `requirements-common.txt` ?

Reply:
@youkaichao 

> is VLLM_TARGET_DEVICE="xpu" pip install git+https://github.com/vllm-project/vllm supported?

Well, I'd argue it should be: for example we're already using the PEP-517 style build syntax in  `Dockerfile.neuron`:
https://github.com/vllm-project/vllm/blob/main/Dockerfile.neuron#L34-L36

We support it by defining the `build-system` section and `requires` list in `pyproject.toml` (see [PEP-517](https://peps.python.org/pep-0517) and [PEP-518](https://peps.python.org/pep-0518))

> isn't it just adding to requirements-common.txt ?

I'm not sure what you mean here, but we have to differentiate between build-time dependencies and runtime dependencies.

Reply:
> Well, I'd argue it should be

I don't see the value. In particular, doing string search / replacement in dockerfiles makes later pytorch upgrade more complicated.

Installing vllm on other platforms is only meant for advanced users, and they should strictly follow the installation guide. They shouldn't expect this works as smoothly as the nvidia platform.

Reply:
@youkaichao 

> I don't see the value. In particular, doing string search / replacement in dockerfiles makes later pytorch upgrade more complicated.

In most cases, we currently build using `python setup.py bdist_wheel`, which is deprecated see ( https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html and https://packaging.python.org/en/latest/discussions/setup-py-deprecated/)

```bash
python setup.py bdist_wheel
```
> ...
> vllm/.venv/lib/python3.11/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
>!!
>
>        ********************************************************************************
>        Please avoid running ``setup.py`` directly.
>        Instead, use pypa/build, pypa/installer or other
>        standards-based tools.
>
>        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
>        ********************************************************************************
> ....

Hence we'll eventually have to move on from this. 

My points for [moving over to a fully PEP518 compliant build system](https://github.com/vllm-project/vllm/pull/7525) include:

- moving over from a deprecated build system
- fixing build-time dependency resolution: the current approach is broken and fragmented, as explained above
- unifying build time dependency resolution in a single file, simplifying dependency management
- avoiding using different build mechanisms for different platforms (e.g. PEP518 builds vs deprecated `setup.py bdist_wheel`
- simplifying the build/install process for different platforms 
(which should help with development as well).

>I don't see the value. In particular, doing string search / replacement in dockerfiles makes later pytorch upgrade more complicated.

That's why I would like to push forward with https://github.com/vllm-project/vllm/pull/7525.

Reply:
I agree it is good to be PEP518 compliant, but I don't think we should do it in this PR, especially the string search / replacement in dockerfiles.

I like the idea in #7525 , maybe you can proceed there.

for this pr to proceed, I suggest you only change the cuda side files.

Reply:
In the latest push, I removed all of the `sed` replacements, replacing that with a `pip install cmake>=3.26 ninja packaging setuptools-scm>=8 wheel jinja2`, the difference with `requirements-build.txt` being the missing `torch` dependency, which we assume already being installed somehow.